other/word_search.py

```diff
@@ -0,0 +1,395 @@
+                letters_left = [self.board[row][col - i] for i in range(word_length)]
+                if all(letter is None for letter in letters_left):
+                    # Successful, insert the word north
+                    for i in range(word_length):
+                        self.board[row][col - i] = word[i]
+                    return
+
+    def insert_north_west(self, word: str, rows: list[int], cols: list[int]) -> None:
+        """
+        >>> ws = WordSearch(WORDS, 3, 3)
+        >>> ws.insert_north_west("cat", [2], [2])
+        >>> ws.board  # doctest: +NORMALIZE_WHITESPACE
+        [['t', None, None],
+        [None, 'a', None],
+        [None, None, 'c']]
+        >>> ws.insert_north_west("at", [1, 2], [0, 1])
+        >>> ws.board  # doctest: +NORMALIZE_WHITESPACE
+        [['t', None, None],
+        ['t', 'a', None],
+        [None, 'a', 'c']]
+        """
+        word_length = len(word)
+        # Attempt to insert the word into each row and when successful, exit
+        for row in rows:
+            # Check if there is space for the word above the row
+            if word_length > row + 1:
+                continue
+
+            # Attempt to insert the word into each column
+            for col in cols:
+                # Check if there is space to the right of the word as well as above
+                if word_length > col + 1:
+                    continue
+
+                # Check if there are existing letters
+                # to the right of the column that will be overwritten
+                letters_diagonal_left = [
+                    self.board[row - i][col - i] for i in range(word_length)
+                ]
+                if all(letter is None for letter in letters_diagonal_left):
+                    # Successful, insert the word north
+                    for i in range(word_length):
+                        self.board[row - i][col - i] = word[i]
+                    return
+
+    def generate_board(self) -> None:
+        """
+        Generates a board with a random direction for each word.
+
+        >>> wt = WordSearch(WORDS, WIDTH, HEIGHT)
+        >>> wt.generate_board()
+        >>> len(list(filter(lambda word: word is not None, sum(wt.board, start=[])))
+        ... ) == sum(map(lambda word: len(word), WORDS))
+        True
+        """
+        directions = (
+            self.insert_north,
+            self.insert_north_east,
+            self.insert_east,
+            self.insert_south_east,
+            self.insert_south,
+            self.insert_south_west,
+            self.insert_west,
+        )
+        for word in self.words:
+            # Shuffle the row order and column order that is used when brute forcing
+            # the insertion of the word
+            rows, cols = list(range(self.height)), list(range(self.width))
+            shuffle(rows)
+            shuffle(cols)
+
+            # Insert the word via the direction
+            choice(directions)(word, rows, cols)
```

Comment:
Isn't it possible for the `insert_...` functions to fail to insert a word? Can we ensure that the word will actually be inserted?

Reply:
You are correct, the function will fail if there is not enough space on the board to insert the word. I will refactor the logic of the insertions to return whether or not the insertion was successful or not, and then raise and error if the insertion was unsuccessful